core/trino-main/src/main/java/io/trino/sql/ir/optimizer/rule/SimplifyContinuousInValues.java

```diff
@@ -0,0 +1,125 @@
+import io.trino.spi.type.DateType;
+import io.trino.spi.type.DecimalType;
+import io.trino.spi.type.IntegerType;
+import io.trino.spi.type.SmallintType;
+import io.trino.spi.type.TimeType;
+import io.trino.spi.type.TimestampType;
+import io.trino.spi.type.TinyintType;
+import io.trino.spi.type.Type;
+import io.trino.sql.ir.Between;
+import io.trino.sql.ir.Constant;
+import io.trino.sql.ir.Expression;
+import io.trino.sql.ir.In;
+import io.trino.sql.ir.IsNull;
+import io.trino.sql.ir.optimizer.IrOptimizerRule;
+import io.trino.sql.planner.Symbol;
+
+import java.math.BigInteger;
+import java.util.Map;
+import java.util.Optional;
+
+import static io.trino.sql.ir.IrUtils.or;
+
+/**
+ * Simplify IN expression with continuous range of constant test values into a BETWEEN expression. E.g,
+ * <ul>
+ *     <li>{@code $in(x, [1, 2, 3, 4]) -> $between(x, 1, 4)}
+ * </ul>
+ */
+public class SimplifyContinuousInValues
+        implements IrOptimizerRule
+{
+    @Override
+    public Optional<Expression> apply(Expression expression, Session session, Map<Symbol, Expression> bindings)
+    {
+        if (!(expression instanceof In in)) {
+            return Optional.empty();
+        }
+
+        if (in.valueList().size() < 2) {
+            return Optional.empty();
+        }
+
+        Type valueType = in.value().type();
+        if (!isDirectLongComparisonValidType(valueType)) {
+            return Optional.empty();
+        }
+
+        if (valueType.getJavaType() != long.class) {
+            return Optional.empty();
+        }
+
+        boolean nullMatch = false;
+        long nonNullsCount = 0;
+        long min = Long.MAX_VALUE;
+        long max = Long.MIN_VALUE;
+        for (Expression testExpression : in.valueList()) {
+            if (!(testExpression instanceof Constant constant)) {
+                return Optional.empty();
+            }
+            if (constant.value() == null) {
+                nullMatch = true;
+                continue;
+            }
+            long longConstant = (long) constant.value();
+            min = Math.min(min, longConstant);
+            max = Math.max(max, longConstant);
+            nonNullsCount++;
+        }
+
+        // If all values within a range are included, use a range filter
+        if (nonNullsCount >= 2 && areAllValuesInRangeIncluded(max, min, nonNullsCount)) {
+            Between between = new Between(in.value(), new Constant(valueType, min), new Constant(valueType, max));
+            if (nullMatch) {
+                return Optional.of(or(new IsNull(in.value()), between));
+            }
+            return Optional.of(between);
+        }
+
+        return Optional.empty();
+    }
+
+    private static boolean isDirectLongComparisonValidType(Type type)
+    {
+        // Types for which we can safely use equality and comparison on the stored long value
+        // instead of going through type specific methods
+        return type instanceof TinyintType ||
```

Comment:
Forgot to mention it on the last round, but we should have a test for each of these types. The primary concern is whether the underlying assumption might change, rendering this check invalid for some types.

Reply:
Added a test verifying type behaviour